### PR TITLE
Feature/transactions wallet v3

### PR DESCRIPTION
**Pull request by giovanni-aptoide**
**What does this PR do?**

   Transactions List with home preview using compose, new transactions api and coroutines.
* Pagination is gonna be done on the transaction details PR

**Database changed?**

   Yes | No

**Where should the reviewer start?**

- [ ] Foo.java
- [ ] Foo.kt

**How should this be manually tested?**

  Flow on how to test this.

**What are the relevant tickets?**

  Tickets related to this pull-request: [APPC-3586](https://aptoide.atlassian.net/browse/APPC-3586)

**Code Review Checklist**

- [ ] Architecture
- [ ] Documentation on public interfaces
- [ ] Database changed?
- [ ] Database migration (if applicable)?
- [ ] Remove comments & unused code & forgotten testing Logs
- [ ] Codestyle
- [ ] Functional QA tests pass

### CODE REVIEW

File: core/network/backend/src/main/java/com/appcoins/wallet/core/network/backend/HandleApi.kt [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package com.appcoins.wallet.core.network.backend
```
Comment (danfbfern) [inline core/network/backend/src/main/java/com/appcoins/wallet/core/network/backend/HandleApi.kt:1]: assuming that this could be here to stay, even though we need to discuss it further, shouldnt this be in the network base module instead of the backend module?

File: ui/widgets/src/main/java/com/appcoins/wallet/ui/widgets/TransactionsComposable.kt [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+  Box(contentAlignment = Alignment.BottomEnd) {
+    if (icon != null) Icon(
+      painter = painterResource(id = icon),
+      contentDescription = null,
+      tint = Color.Unspecified,
+      modifier = Modifier.size(40.dp)
+    ) else {
+      SubcomposeAsyncImage(
+        model = appIcon,
+        contentDescription = "Game Icon",
```
Comment (danfbfern) [inline ui/widgets/src/main/java/com/appcoins/wallet/ui/widgets/TransactionsComposable.kt:118]: hardcoded description might be an issue in alternative languages

File: app/src/main/java/com/asfoundation/wallet/transactions/TransactionsListViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+      defaultCurrency = selectedCurrency
+      fetchTransactions(wallet.address, selectedCurrency)
+    }.subscribe()
+  }
+
+  private fun fetchTransactions(walletAddress: String, selectedCurrency: String) {
+    viewModelScope.launch {
+      fetchTransactionsHistoryUseCase(wallet = walletAddress, currency = selectedCurrency)
+        .onStart { _uiState.value = UiState.Loading }
+        .catch { logger.log(TAG, it) }
```
Comment (danfbfern) [inline app/src/main/java/com/asfoundation/wallet/transactions/TransactionsListViewModel.kt:59]: when catching shouldn’t we be handling it and wraping the error? otherwise what is the point of trying to have a Result being mapped since the repository?

File: app/src/main/java/com/asfoundation/wallet/wallet/home/HomeViewModel.kt [hunk 4]
```diff
@@ -229,7 +256,7 @@ class HomeViewModel @Inject constructor(
     if (walletModel == null) return Observable.empty()
     val retainValue = if (walletModel.isNewWallet) null else HomeState::transactionsModelAsync
     return Observable.combineLatest(
-      getTransactions(walletModel.wallet), getCardNotifications(),
+      Observable.just(listOf()), getCardNotifications(),
```
Comment (danfbfern) [inline app/src/main/java/com/asfoundation/wallet/wallet/home/HomeViewModel.kt:259]: What is the purpose of this change?

File: core/network/backend/src/main/java/com/appcoins/wallet/core/network/backend/remote/RemoteTransactionDataSource.kt [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+class RemoteTransactionDataSource @Inject constructor(private val api: TransactionsApi) :
+  BackendDataSource {
+
+  override suspend fun getTransactionsHistory(
+    wallet: String,
+    selectedCurrency: String,
+    limit: Int
+  ) = api.getTransactionHistory(wallet = wallet, defaultCurrency = selectedCurrency, limit = limit)
+}
```
Comment (danfbfern) [inline core/network/backend/src/main/java/com/appcoins/wallet/core/network/backend/remote/RemoteTransactionDataSource.kt:15]: Isn’t this simply an unecessary abstraction?
The api could be called direclty in the repository instead of having this codesmell (middle man) doing only the job of calling the api anyway.
Combined with its interface, we have 2 classes plus 2 interfaces that could be simply one repository since we aren’t using the repo interface for more than one implements

**Review by danfbfern — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/asfoundation/wallet/repository/TransactionsHistoryRepository.kt [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+import kotlinx.coroutines.flow.flowOn
+import javax.inject.Inject
+
+interface TransactionsHistoryRepository {
+  suspend fun fetchTransactions(
+    wallet: String,
+    limit: Int,
+    currency: String
+  ): Flow<ApiResult<List<TransactionResponse>>>
+}
```
[the remark this answers is not in the document]
Comment (giovanni-aptoide) [reply app/src/main/java/com/asfoundation/wallet/repository/TransactionsHistoryRepository.kt:20]: Yes, it's gonna be used to make the FakeRepository to fake tests,

File: core/network/backend/src/main/java/com/appcoins/wallet/core/network/backend/HandleApi.kt (hunk 1 above)
Comment (giovanni-aptoide) [reply core/network/backend/src/main/java/com/appcoins/wallet/core/network/backend/HandleApi.kt:1]: Yes, I will move it

File: ui/widgets/src/main/java/com/appcoins/wallet/ui/widgets/TransactionsComposable.kt (hunk 2 above)
Comment (giovanni-aptoide) [reply ui/widgets/src/main/java/com/appcoins/wallet/ui/widgets/TransactionsComposable.kt:118]: you're right, my mistake, I will change it

File: app/src/main/java/com/asfoundation/wallet/transactions/TransactionsListViewModel.kt (hunk 3 above)
Comment (giovanni-aptoide) [reply app/src/main/java/com/asfoundation/wallet/transactions/TransactionsListViewModel.kt:59]: I'm just logging the error because we don't have other way to handle errors defined, like dialogs or something

File: app/src/main/java/com/asfoundation/wallet/wallet/home/HomeViewModel.kt (hunk 4 above)
Comment (giovanni-aptoide) [reply app/src/main/java/com/asfoundation/wallet/wallet/home/HomeViewModel.kt:259]: It was just to stop fetching the transactions from the another api? I can remove it totally, this approach was just to avoid bugs in other flows

File: core/network/backend/src/main/java/com/appcoins/wallet/core/network/backend/remote/RemoteTransactionDataSource.kt (hunk 5 above)
Comment (giovanni-aptoide) [reply core/network/backend/src/main/java/com/appcoins/wallet/core/network/backend/remote/RemoteTransactionDataSource.kt:15]: I don’t think so, suppose I need to get some data from the local database and some data from network, don’t you prefer to have that separatedly?

Comment (danfbfern) [reply core/network/backend/src/main/java/com/appcoins/wallet/core/network/backend/remote/RemoteTransactionDataSource.kt:15]: I get that, but to have a class do simply call the api and return that, acheives nothing other than adding complexity to the code.
If the Repository need to get data from the network and from the database, the repository should call the network and database instead of having a class and interface just to do this ‘api.getTransactionHistory(wallet = wallet, defaultCurrency = selectedCurrency, limit = limit)’.
Calling the api directly or calling this dataSource its the same in terms of separation but this implemetantion adds complexity because it adds a middle man class and adds an interface that will always only have one implementation

File: app/src/main/java/com/asfoundation/wallet/wallet/home/HomeViewModel.kt (hunk 4 above)
Comment (danfbfern) [reply app/src/main/java/com/asfoundation/wallet/wallet/home/HomeViewModel.kt:259]: It looks weird to have a Observable to do nothing, it created unecessary memory allocation (assuming that the previous code is not needed to get the transactions anymore).
If its not need, I would say to simply remove it and since this is inside the HomeViewModel, im assuming no other flow will break because of it.
Especially because passing this empty list as transactions to other methods might generate issues in debug because it looks like that might return the transactions and in this case the TransactionsModel is the only think affected and it appears to be used in the old flow home, but in this branch the old home should be removed anyways, so we can even remove this TransactionsModel if its not being used

Comment (giovanni-aptoide) [reply app/src/main/java/com/asfoundation/wallet/wallet/home/HomeViewModel.kt:259]: I got it, already done!

File: core/network/backend/src/main/java/com/appcoins/wallet/core/network/backend/remote/RemoteTransactionDataSource.kt (hunk 5 above)
Comment (giovanni-aptoide) [reply core/network/backend/src/main/java/com/appcoins/wallet/core/network/backend/remote/RemoteTransactionDataSource.kt:15]: I agree with you, done!

**Review by gustavoaptoide — APPROVED** [review]
